tokio-util/src/task/task_tracker.rs

```diff
@@ -0,0 +1,629 @@
+    #[inline]
+    #[track_caller]
+    #[cfg(feature = "rt")]
+    #[cfg(not(target_family = "wasm"))]
+    #[cfg_attr(docsrs, doc(cfg(feature = "rt")))]
+    pub fn spawn_blocking<F, T>(&self, task: F) -> JoinHandle<T>
+    where
+        F: FnOnce() -> T,
+        F: Send + 'static,
+        T: Send + 'static,
+    {
+        let token = self.token();
+        tokio::task::spawn_blocking(move || {
+            let res = task();
+            drop(token);
+            res
+        })
+    }
+
+    /// Spawn the provided blocking task on the provided Tokio runtime, and track it in this `TaskTracker`.
+    ///
+    /// This is equivalent to `handle.spawn_blocking(tracker.track_future(task))`.
+    #[inline]
+    #[track_caller]
+    #[cfg(feature = "rt")]
+    #[cfg(not(target_family = "wasm"))]
+    #[cfg_attr(docsrs, doc(cfg(feature = "rt")))]
+    pub fn spawn_blocking_on<F, T>(&self, task: F, handle: &Handle) -> JoinHandle<T>
+    where
+        F: FnOnce() -> T,
+        F: Send + 'static,
+        T: Send + 'static,
+    {
+        let token = self.token();
+        handle.spawn_blocking(move || {
+            let res = task();
+            drop(token);
+            res
+        })
+    }
+
+    /// Track the provided future.
+    ///
+    /// The returned [`TrackedFuture`] will count as a task tracked by this collection, and will
+    /// prevent calls to [`wait`] from returning.
+    ///
+    /// The task is removed from the collection when it is dropped. (It is not removed immediately
+    /// when [`poll`] returns [`Poll::Pending`]. You have to actually run the destructor for it to
+    /// be removed.)
+    ///
+    /// # Examples
+    ///
+    /// Track a future spawned on a [`JoinSet`].
+    /// ```
+    /// # async fn my_async_fn() {}
+    /// use tokio::task::JoinSet;
+    /// use tokio_util::task::TaskTracker;
+    ///
+    /// # #[tokio::main(flavor = "current_thread")]
+    /// # async fn main() {
+    /// let tracker = TaskTracker::new();
+    /// let mut join_set = JoinSet::new();
+    ///
+    /// join_set.spawn(tracker.track_future(my_async_fn()));
+    /// # }
+    /// ```
+    ///
+    /// [`JoinSet`]: tokio::task::JoinSet
+    /// [`Poll::Pending`]: std::task::Poll::Pending
+    /// [`poll`]: std::future::Future::poll
+    /// [`wait`]: Self::wait
+    #[inline]
+    pub fn track_future<F: Future>(&self, future: F) -> TrackedFuture<F> {
+        TrackedFuture {
+            future,
+            token: self.token(),
+        }
+    }
+
+    /// Creates a [`TaskTrackerToken`] representing a task tracked by this `TaskTracker`.
+    ///
+    /// This token is a lower-level utility than the spawn methods. Each token is considered to
+    /// correspond to a task. As long as the token exists, the `TaskTracker` cannot complete.
+    /// Furthermore, the count returned by the [`len`] method will include the tokens in the count.
+    ///
+    /// Dropping the token corresponds to when the task exits.
```

Comment:
This wording feels a little bit awkward, what do you think about something like this:
```suggestion
    /// Dropping the token indicates to the `TaskTracker` that the task has exited.
```